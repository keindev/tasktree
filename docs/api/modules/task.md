# Module: Task

## Table of contents

### Enumerations

- [TaskStatus](../enums/task.taskstatus.md)

### Classes

- [Task](../classes/task.task-1.md)

### Interfaces

- [ITaskOptions](../interfaces/task.itaskoptions.md)
