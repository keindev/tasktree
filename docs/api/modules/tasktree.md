# Module: TaskTree

## Table of contents

### Enumerations

- [ExitCode](../enums/tasktree.exitcode.md)

### Classes

- [TaskTree](../classes/tasktree.tasktree-1.md)

### Interfaces

- [ITaskTreeOptions](../interfaces/tasktree.itasktreeoptions.md)
