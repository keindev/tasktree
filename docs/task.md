# Task

Entity for managing a task - includes all child objects (informational messages, errors, progress bars, and tasks).

## Task statuses

-   Pending: `0`
-   Completed: `1`
-   Failed: `2`
-   Skipped: `3`

## API

### id()

Returns task `uid`.

### getText()

Returns task text to display.

### getStatus()

Returns current task [status](#task-statuses).

### getActive()

Returns the first leaf subtask of the task tree, if it exists, otherwise, the object itself.

### isPending()

Returns a Boolean value - indicating the [status](#task-statuses) of the task.

### haveWarnings()

Returns a Boolean value - indicating the [task](#task) have warnings.

### haveErrors()

Returns a Boolean value - indicating the [task](#task) have errors.

### havePendingSubtasks()

Returns a Boolean value - indicating the [status](#task-statuses) of the subtask.

### haveSubtasks()

Returns a boolean value - indicating the presence of subtasks.

### add(text \[, options\])

Adds a new subtask. Returns a [subtask](#task) object.

#### text

Text for display.

Type: `string`

#### Options

New subtask options.

##### status

Type: `number`

Default: `0`

New subtask [status](#task-statuses).

##### autoClear

Type: `boolean`

Default: `false`

Removes all subtasks and progress bars after complete.

### update(text)

Update task text.

#### text

Text for display.

Type: `string`

### bar(\[template\, options\])

Adds a new progress bar. Returns a [progress bar](#progress-bar) object.

#### template

Type: `string`

Default: `:bar :rate/bps :percent :etas`

Text for display with [tokens](#progress-bar-tokens).

#### options

Type: `object`

Default: `null`

These are keys in the options object you can pass to the progress bar along. See [ProgressBar.Options](./progress-bar.md)

### clear()

Removes all subtasks and progress bars.

### complete(\[text, clear\])

Complete task. Returns [self-object](#task).

#### text

Type: `string`

Default: `''`

Text for display. Modifies task text if submitted.

#### clear

Type: `boolean`

Default: `false`

Calls `clear()` if true.

#### skip(\[text, clear\])

Skip task. Returns [self-object](#task).

#### text

Type: `string`

Default: `''`

Text for display. Modifies task text if submitted.

#### clear

Type: `boolean`

Default: `false`

Calls `clear()` if true.

### fail(\[text, clear\])

Failure task. Throws exception in "silent mode".

#### text

Type: `string`

Default: `''`

Text for display. Modifies task text if submitted.

#### clear

Type: `boolean`

Default: `false`

Calls `clear()` if true.

### error(error \[, fail\])

Adds an error message to the task, which will be displayed immediately under the task header. Returns [self-object](#task).

#### error

Type: `string | Error`

Error message.

#### fail

Type: `boolean`

Default: `false`

Flag indicating the need to call the `fail` method.

### log(\[text\])

Adds an informational message, to be displayed under the task title. Returns [self-object](#task).

#### text

Type: `string`

Informational message.

### warn(\[text\])

Adds a warning message, to be displayed under the task title. Returns [self-object](#task).

#### text

Type: `string`

Warning message.
