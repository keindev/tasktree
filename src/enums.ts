export enum Status {
    Pending = 0,
    Completed = 1,
    Failed = 2,
    Skipped = 3,
}

export enum Level {
    Default = 0,
    Step = 1,
    Stride = 3,
}

export enum Progress {
    Default = -1,
    Start = 0,
    End = 100,
}

export enum Token {
    // the progress bar itself
    Bar = ':bar',
    // current tick number
    Current = ':current',
    // total ticks
    Total = ':total',
    // time elapsed in seconds
    Elapsed = ':elapsed',
    // completion percentage
    Percent = ':percent',
    // estimated completion time in seconds
    ETA = ':eta',
    // rate of ticks per second
    Rate = ':rate',
}

export enum Indent {
    Default = 2,
    Empty = 0,
}

export enum Type {
    Default = 'default',
    Active = 'active',
    Success = 'success',
    Skip = 'skip',
    Error = 'error',
    Message = 'message',
    Info = 'info',
    Warning = 'warning',
    Subtask = 'subtask',
    List = 'list',
    Dim = 'dim',
}

export enum Color {
    Default = '',
    Active = '#4285f4',
    Success = '#00c851',
    Skip = '#ff8800',
    Error = '#ff4444',
    Message = '#2e2e2e',
    Info = '#33b5e5',
    Warning = '#ffbb33',
    Subtask = '#2e2e2e',
    List = '#4285f4',
    Dim = '#838584',
}

export enum Badge {
    Default = '',
    Error = '[fail]',
    Skip = '[skip]',
}
