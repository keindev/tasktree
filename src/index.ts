export { default } from './TaskTree.js';
