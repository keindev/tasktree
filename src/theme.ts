import chalk from 'chalk';
import convert from 'color-convert';
import elegantSpinner from 'elegant-spinner';
import * as Figures from 'figures';
import * as Types from './types';
import * as Enums from './enums';
import { Task } from './task';
import { Progress } from './progress';

const frame = elegantSpinner();

export class Theme {
    public static SEPARATOR = '\n';
    public static EMPTY = '';
    public static SPACE = ' ';
    public static INDENT = '  ';

    private colors: Map<Enums.Type, Types.Color> = new Map();
    private figures: Map<Enums.Type, Types.Figure> = new Map();
    private badges: Map<Enums.Type, Types.Badge> = new Map();

    public constructor(options?: Types.Theme) {
        if (options) {
            const { colors, figures, badges } = this;
            const list = new Map(Object.entries(options));
            const set = (key: Enums.Type, color?: Types.Color, figure?: Types.Figure, badge?: Types.Badge): void => {
                if (color) colors.set(key, color);
                if (figure) figures.set(key, figure);
                if (badge) badges.set(key, badge);
            };

            Object.values(Enums.Type).forEach(
                (key: Enums.Type): void => {
                    const option = list.get(key);

                    if (Array.isArray(option)) {
                        set(key, ...option);
                    } else if (typeof option === 'object') {
                        const { color, figure, badge } = option;

                        set(key, color, figure, badge);
                    }
                }
            );
        }
    }

    public static type(status: Enums.Status, isList: boolean = false): Enums.Type {
        let type: Enums.Type;

        switch (status) {
            case Enums.Status.Completed:
                type = Enums.Type.Success;
                break;
            case Enums.Status.Skipped:
                type = Enums.Type.Skip;
                break;
            case Enums.Status.Failed:
                type = Enums.Type.Error;
                break;
            case Enums.Status.Pending:
            default:
                type = isList ? Enums.Type.List : Enums.Type.Active;
                break;
        }

        return type;
    }

    public static join(separator: string, ...text: string[]): string {
        return text.filter((value): boolean => !!value.length).join(separator);
    }

    public static dye(str: string, color: Types.Color): string {
        return color ? chalk.hex(color)(str) : str;
    }

    public static indent(count: number, ...text: string[]): string {
        return `${Theme.INDENT.padStart(count * Enums.Indent.Default)}${Theme.join(Theme.SPACE, ...text)}`;
    }

    private static getValueBy<T>(map: Map<Enums.Type, T>, type: Enums.Type, getDefault: () => T): T {
        let result = map.get(type);

        if (!result) result = getDefault();

        return result;
    }

    public getColor(type: Enums.Type): Types.Color {
        const { colors } = this;

        return Theme.getValueBy(
            this.colors,
            type,
            (): string => {
                if (type === Enums.Type.Active) return Enums.Color.Active;
                if (type === Enums.Type.Success) return Enums.Color.Success;
                if (type === Enums.Type.Skip) return Enums.Color.Skip;
                if (type === Enums.Type.Error) return Enums.Color.Error;
                if (type === Enums.Type.Message) return Enums.Color.Message;
                if (type === Enums.Type.Info) return Enums.Color.Info;
                if (type === Enums.Type.Warning) return Enums.Color.Warning;
                if (type === Enums.Type.Exception) return Enums.Color.Exception;
                if (type === Enums.Type.Subtask) return Enums.Color.Subtask;
                if (type === Enums.Type.List) return Enums.Color.List;
                if (type === Enums.Type.Dim) return Enums.Color.Dim;

                return colors.get(Enums.Type.Default) || Enums.Color.Default;
            }
        );
    }

    public paint(str: string, type: Enums.Type): string {
        return Theme.dye(str, this.getColor(type));
    }

    public gradient(str: string, gradient: Types.Gradient): string {
        const begin = convert.hex.rgb(this.getColor(gradient.begin));
        const end = convert.hex.rgb(this.getColor(gradient.end));
        const w = gradient.position * 2 - 1;
        const w1 = (w + 1) / 2.0;
        const w2 = 1 - w1;

        return Theme.dye(
            str,
            convert.rgb.hex([
                Math.round(end[0] * w1 + begin[0] * w2),
                Math.round(end[1] * w1 + begin[1] * w2),
                Math.round(end[2] * w1 + begin[2] * w2),
            ])
        );
    }

    public figure(type: Enums.Type): Types.Figure {
        const { figures } = this;
        const figure = Theme.getValueBy(
            figures,
            type,
            (): string => {
                if (type === Enums.Type.Active) return frame();
                if (type === Enums.Type.Success) return Figures.tick;
                if (type === Enums.Type.Skip) return Figures.arrowDown;
                if (type === Enums.Type.Error) return Figures.cross;
                if (type === Enums.Type.Message) return Figures.line;
                if (type === Enums.Type.Info) return Figures.info;
                if (type === Enums.Type.Warning) return Figures.warning;
                if (type === Enums.Type.Exception) return Figures.arrowRight;
                if (type === Enums.Type.Subtask) return Figures.pointerSmall;
                if (type === Enums.Type.List) return Figures.pointer;

                return figures.get(Enums.Type.Default) || this.figure(Enums.Type.Subtask);
            }
        );

        return figure ? this.paint(figure, type) : figure;
    }

    public badge(type: Enums.Type): Types.Badge {
        const { badges } = this;
        const badge = Theme.getValueBy(
            badges,
            type,
            (): string => {
                if (type === Enums.Type.Error) return Enums.Badge.Error;
                if (type === Enums.Type.Skip) return Enums.Badge.Skip;

                return badges.get(Enums.Type.Default) || Enums.Badge.Default;
            }
        );

        return badge ? this.paint(badge, Enums.Type.Dim) : badge;
    }

    public title(task: Task, level: number): string {
        const type = Theme.type(task.getStatus(), task.isList());
        const badge = this.badge(type);
        const figure = this.figure(type);
        let prefix = Theme.EMPTY;

        if (level) prefix = task.isList() ? this.figure(Enums.Type.Subtask) : this.figure(Enums.Type.Default);

        return Theme.indent(level, prefix, figure, task.getText(), badge);
    }

    public errors(errors: string[], level: number): string[] {
        const type = Enums.Type.Error;
        const sublevel = level + Enums.Level.Stride;

        return errors.map(
            (text): string => {
                const [error, ...lines] = text.split(Theme.SEPARATOR);
                const title = Theme.join(Theme.SPACE, this.figure(Enums.Type.Message), this.figure(type), error.trim());

                return Theme.join(
                    Theme.SEPARATOR,
                    Theme.indent(level + Enums.Level.Step, this.paint(title, type)),
                    ...lines.map((line): string => Theme.indent(sublevel, this.paint(line.trim(), Enums.Type.Dim)))
                );
            }
        );
    }

    public messages(list: string[], type: Enums.Type, level: number): string[] {
        const figure = this.figure(type);
        const sign = this.figure(Enums.Type.Message);
        const indent = level + Enums.Level.Step;

        return list.map((text): string => Theme.indent(indent, sign, figure, text));
    }

    public bars(list: Progress[], level: number): string[] {
        const figure = this.figure(Enums.Type.Subtask);

        return list
            .filter((bar): boolean => !bar.isCompleted() || !bar.clear)
            .map((bar): string => Theme.indent(level, figure, bar.render(this)));
    }
}
