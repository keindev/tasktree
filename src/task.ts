import chalk from 'chalk';
import { TaskTree } from './tasktree';
import { Theme } from './theme';
import * as Enums from './enums';
import { Progress } from './progress';
import { Options } from './types';

let uid = 0;

export class Task {
    private uid: number;
    private text: string;
    private status: Enums.Status;
    private bars: Progress[] = [];
    private subtasks: Task[] = [];
    private logs: Set<string> = new Set();
    private errors: string[] = [];
    private warnings: Set<string> = new Set();

    public constructor(text: string, status: Enums.Status = Enums.Status.Pending) {
        this.uid = ++uid;
        this.text = text;
        this.status = status;
    }

    public id(): number {
        return this.uid;
    }

    public getText(): string {
        return this.text;
    }

    public getStatus(): Enums.Status {
        return this.status;
    }

    public getActive(): Task {
        const { subtasks } = this;
        const subtask = subtasks[subtasks.length - 1];
        let task: Task = this;

        if (subtask && subtask.isPending()) task = subtask.getActive();

        return task;
    }

    public isPending(): boolean {
        return this.status === Enums.Status.Pending;
    }

    public havePendingSubtasks(): boolean {
        return !!this.subtasks.filter((task): boolean => task.isPending()).length;
    }

    public haveSubtasks(): boolean {
        return !!this.subtasks.length;
    }

    public add(text: string, status: Enums.Status = Enums.Status.Pending): Task {
        const isCompleted = !this.isPending();
        const task = new Task(text, isCompleted ? Enums.Status.Failed : status);

        this.subtasks.push(task);

        if (isCompleted) this.fail(`Task is already complete, can't add new subtask [${task.id()}]`);

        return task;
    }

    public bar(template?: string, options?: Options): Progress {
        const isCompleted = !this.isPending();
        const bar = new Progress(template, isCompleted ? { total: Enums.Progress.End } : options);

        this.bars.push(bar);

        if (isCompleted) this.fail("Task is already complete, can't add new progress bar");

        return bar;
    }

    public clear(): void {
        this.subtasks = [];
        this.bars = [];
    }

    public complete(text?: string, clear: boolean = false): Task {
        if (this.havePendingSubtasks()) this.fail('Subtasks is not complete.');

        this.update(Enums.Status.Completed, text, clear);
        this.bars = this.bars.filter((bar): boolean => {
            bar.complete();

            return !bar.clear;
        });

        return this;
    }

    public skip(text?: string, clear: boolean = false): Task {
        this.update(Enums.Status.Skipped, text, clear);

        return this;
    }

    public fail(text?: string, clear: boolean = false): never {
        this.update(Enums.Status.Failed, text, clear);

        TaskTree.tree().exit(Enums.ExitCode.Error);

        throw new Error(text);
    }

    public error(error?: string | Error, fail?: boolean): Task {
        const { errors } = this;

        if (typeof error === 'string') errors.push(error);
        if (error instanceof Error && error.stack) errors.push(error.stack);
        if (fail) this.fail();

        return this;
    }

    public log(text: string): Task {
        if (this.isPending()) this.logs.add(text);

        return this;
    }

    public warn(text: string): Task {
        if (this.isPending()) this.warnings.add(text);

        return this;
    }

    public render(theme: Theme, level = Enums.Level.Default): string[] {
        const type = level ? Enums.Type.Dim : Enums.Type.Default;
        const rows = [
            theme.title(this, level),
            ...theme.bars(this.bars, level + Enums.Level.Step),
            ...theme.errors(this.errors, level),
            ...theme.messages([...this.warnings], Enums.Type.Warning, level),
            ...theme.messages([...this.logs], Enums.Type.Info, level),
        ];

        this.subtasks.forEach((task): void => {
            rows.push(...task.render(theme, level + Enums.Level.Step));
        });

        return rows.map((row): string => theme.paint(row, type));
    }

    private update(status: Enums.Status, text?: string, clear: boolean = false): void {
        if (this.isPending()) {
            if (text) this.text = text;
            if (clear) this.clear();

            this.status = status;
        } else {
            this.error(`Task is already complete (${chalk.bold(this.status.toString())})`);
            this.status = Enums.Status.Failed;
        }
    }
}
