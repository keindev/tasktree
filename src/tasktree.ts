import { UpdateManager } from 'stdout-update';
import { Task } from './task';
import * as Types from './types';
import { Theme } from './theme';
import { ExitCode } from './enums';

export class TaskTree {
    public static TIMEOUT = 100;
    private static instance: TaskTree;

    private handle: NodeJS.Timeout | undefined;
    private tasks: Task[];
    private theme: Theme;
    private manager: UpdateManager;
    private silence: boolean = false;

    private constructor(theme?: Types.Theme) {
        this.tasks = [];
        this.theme = new Theme(theme);
        this.manager = UpdateManager.getInstance();
    }

    public static tree(theme?: Types.Theme): TaskTree {
        if (!TaskTree.instance) {
            TaskTree.instance = new TaskTree(theme);
        }

        return TaskTree.instance;
    }

    public start(silence?: boolean): TaskTree {
        this.silence = !!silence;
        this.tasks = [];

        if (!this.handle && !this.silence) {
            this.manager.hook();
            this.handle = setInterval((): void => {
                this.log();
            }, TaskTree.TIMEOUT);
        }

        return this;
    }

    public stop(): TaskTree {
        if (this.handle) {
            clearInterval(this.handle);

            this.log();
            this.manager.unhook();
            this.handle = undefined;
        }

        return this;
    }

    public exit(code: ExitCode = ExitCode.Success): void {
        if (!this.silence) {
            this.stop();
            process.exit(code);
        }
    }

    public add(text: string): Task {
        const { tasks } = this;
        let task = tasks[tasks.length - 1];

        if (task && task.isPending()) {
            task = task.getActive();
            task = task.add(text);
        } else {
            tasks.push((task = new Task(text)));
        }

        return task;
    }

    public render(): string[] {
        return this.tasks.reduce<string[]>((acc, task): string[] => acc.concat(task.render(this.theme)), []);
        // return this.tasks.map((task): string[] => task.render(this.theme)).flat();
    }

    private log(): void {
        this.manager.update(this.render());
    }
}
