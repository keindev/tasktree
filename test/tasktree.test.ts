import stripAnsi from 'strip-ansi';
import { Task } from '../src/task';
import { TaskTree } from '../src/tasktree';
import { Theme } from '../src/theme';

const tree = TaskTree.tree();
const theme = new Theme();
const titles = ['Task1', 'Task2', 'Task3'];

let tasks: Task[];

describe('TaskTree', (): void => {
    it('creating', (): void => {
        expect(tree).not.toBeUndefined();

        tree.start(true);
        tasks = titles.map((title): Task => tree.add(title));
    });

    it('manage', (): void => {
        const error = 'Something bad happened\nat X\nat Y\nat Z';
        const result = tasks.reverse().every(
            (task, index): boolean => {
                expect(task.isPending()).toBeTruthy();
                expect(task.render(theme).length).toBeGreaterThan(1);

                task.log(`Log ${index}`).warn(`Warn ${index}`);

                switch (index) {
                    case 0:
                        task.skip();
                        break;
                    case tasks.length - 1:
                        task.error(error).fail();
                        break;
                    default:
                        task.complete();
                        break;
                }

                return !task.isPending();
            }
        );

        expect(result).toBeTruthy();
    });

    it('render', (): void => {
        expect(stripAnsi(tree.render())).toMatchSnapshot();
        tree.stop(true);
        expect(stripAnsi(tree.render())).toMatchSnapshot();

        tree.start(true);
        tree.stop(true);
        expect(stripAnsi(tree.render())).toMatchSnapshot();
    });
});
