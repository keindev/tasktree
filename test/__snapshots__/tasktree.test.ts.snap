// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTree render 1`] = `
"  ✖ Task1: Fail
  ⚠ Warn 2
  ℹ Log 2
  → ✔ Task2
    ⚠ Warn 1
    ℹ Log 1
    → ↓ Task3 [skip]
      ⚠ Warn 0
      ℹ Log 0"
`;

exports[`TaskTree render 2`] = `
"  ✖ Task1: Fail
  ⚠ Warn 2
  ℹ Log 2
  → ✔ Task2
    ⚠ Warn 1
    ℹ Log 1
    → ↓ Task3 [skip]
      ⚠ Warn 0
      ℹ Log 0"
`;

exports[`TaskTree render 3`] = `""`;
